x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/TransactionStore.java

```diff
@@ -0,0 +1,769 @@
+     * Container for holding item ids reverse sorted by count
+     *
+     * Only stores a list of ids, not the count
+     */
+    public static class TopItemIds implements Iterable<Long>, Releasable {
+
+        public class IdIterator implements Iterator<Long> {
+
+            private int currentIndex;
+
+            IdIterator(int startIndex) {
+                currentIndex = startIndex;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return currentIndex < sortedItems.size();
+            }
+
+            @Override
+            public Long next() {
+                return sortedItems.get(currentIndex++);
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+
+            public int getIndex() {
+                return currentIndex;
+            }
+
+            public void reset() {
+                currentIndex = 0;
+            }
+
+            public void reset(int startIndex) {
+                currentIndex = startIndex;
+            }
+        }
+
+        private final LongArray sortedItems;
+
+        TopItemIds(LongArray sortedItems) {
+            this.sortedItems = sortedItems;
+        }
+
+        @Override
+        public IdIterator iterator() {
+            return iterator(0);
+        }
+
+        public IdIterator iterator(int startIndex) {
+            return new IdIterator(startIndex);
+        }
+
+        public long size() {
+            return sortedItems.size();
+        }
+
+        @Override
+        public void close() {
+            Releasables.close(sortedItems);
+        }
+    }
+
+    /**
+     * Container for holding transaction ids reverse sorted by count
+     *
+     * Only stores a list of ids, not the count
+     */
+    public static class TopTransactionIds implements Iterable<Long>, Releasable {
+        private final LongArray sortedTransactions;
+
+        TopTransactionIds(LongArray sortedTransactions) {
+            this.sortedTransactions = sortedTransactions;
+        }
+
+        @Override
+        public Iterator<Long> iterator() {
+            Iterator<Long> it = new Iterator<Long>() {
+
+                private int currentIndex = 0;
+
+                @Override
+                public boolean hasNext() {
+                    return currentIndex < sortedTransactions.size();
+                }
+
+                @Override
+                public Long next() {
+                    return sortedTransactions.get(currentIndex++);
+                }
+
+                @Override
+                public void remove() {
+                    throw new UnsupportedOperationException();
+                }
+            };
+            return it;
+        }
+
+        public long size() {
+            return sortedTransactions.size();
+        }
+
+        @Override
+        public void close() {
+            Releasables.close(sortedTransactions);
+        }
```

Comment:
It seems weird to me that this thing is closing something it didn't even open. Something else creates `sortedTransactions`, but this will close it.

Reply:
it is created by TransactionStore, I change the constructor now to `private`, that should make it more clear. I don't want to move the allocation of the array into the iterators, as it does not seem logical to me.

Reply:
@hendrikmuhs that is fine, the initialization of the arrays doesn't have to be within the iterators, but the thing that closes the arrays probably shouldn't be within the iterators.

This whole transactionstore class has multiple things that end up being closed/released at different times. It smells exceptionally fragile. The contracts on when things get closed or not are not very clear.